internal/engine/upper_test.go

```diff
@@ -3739,7 +3751,9 @@ func (f *testFixture) TearDown() {
 	close(f.fsWatcher.errors)
 	f.b.buildCompletionChans.Range(func(key, value interface{}) bool {
 		if ch, ok := value.(buildCompletionChannel); ok {
-			close(ch)
+			if !ch.isClosed() {
```

Comment:
this still has race conditions, because your check is non-atomic.

tbh i wish we hadn't used channels at all here -- it's not really a good fit for what we're doing.

Reply:
alternatively, just remove this Range() call entirely, it's not really necessary